R/module_snapshot_manager.R

```diff
@@ -0,0 +1,241 @@
+      NULL
+    ),
+    uiOutput(ns("snapshot_list"))
+  )
+}
+
+#' @rdname snapshot_manager_module
+#' @export
+#'
+snapshot_manager_srv <- function(id, slices_global, mapping_matrix, filtered_data_list) {
+  checkmate::assert_character(id)
+  checkmate::assert_true(is.reactive(slices_global))
+  checkmate::assert_class(slices_global(), "teal_slices")
+  checkmate::assert_true(is.reactive(mapping_matrix))
+  checkmate::assert_data_frame(mapping_matrix())
+  checkmate::assert_list(filtered_data_list, types = "FilteredData", any.missing = FALSE, names = "named")
+
+  moduleServer(id, function(input, output, session) {
+
+    ns <- session$ns
+
+    # Store global filter states.
+    filter <- isolate(slices_global())
+    snapshot_history <- reactiveVal({
+      list(
+        "Initial application state" = disassemble_slices(filter)
+      )
+    })
+
+    # Snapshot current application state - name snaphsot.
+    observeEvent(input$snapshot_add, {
+      showModal(
+        modalDialog(
+          textInput(ns("snapshot_name"), "Name the snapshot", width = "100%", placeholder = "Meaningful, unique name"),
+          footer = tagList(
+            actionButton(ns("snapshot_name_accept"), "Accept", icon = icon("thumbs-up")),
+            modalButton(label = "Cancel", icon = icon("thumbs-down"))
+          ),
+          size = "s"
+        )
+      )
+    })
+    # Snapshot current application state - store snaphsot.
+    observeEvent(input$snapshot_name_accept, {
+      snapshot_name <- trimws(input$snapshot_name)
+      if (identical(snapshot_name, "")) {
+        showNotification(
+          "Please name the snapshot.",
+          type = "message"
+        )
+        updateTextInput(inputId = "snapshot_name", value = "", placeholder = "Meaningful, unique name")
+      } else if (is.element(make.names(snapshot_name), make.names(names(snapshot_history())))) {
+        showNotification(
+          "This name is in conflict with other snapshot names. Please choose a different one.",
+          type = "message"
+        )
+        updateTextInput(inputId = "snapshot_name", value = , placeholder = "Meaningful, unique name")
+      } else {
+        snapshot <- disassemble_slices(slices_global())
+        attr(snapshot, "mapping") <- matrix_to_mapping(mapping_matrix())
+        snapshot_update <- c(snapshot_history(), list(snapshot))
+        names(snapshot_update)[length(snapshot_update)] <- snapshot_name
+        snapshot_history(snapshot_update)
+        removeModal()
+        # Reopen filter manager modal by clicking button in the main application.
+        shinyjs::click(id = "teal-main_ui-filter_manager-show", asis = TRUE)
+      }
+    })
+
+    # Restore initial state.
+    observeEvent(input$snapshot_reset, {
+      s <- "Initial application state"
+      ### Begin restore procedure. ###
+      snapshot <- snapshot_history()[[s]]
+      snapshot_state <- reassemble_slices(snapshot)
```

Comment:
reassemble_slices imported from `teal.slice` namespace fails as `teal_slices` has no `mapping` and `module_specific` argument. We need to create a copy in `teal` to use `teal::teal_slices`.

Now we have `store`, `restore`, `as.list`, `slices_to_list`, `as.list`, `reassemble`, `disassemble_slices`. We need to figure out consistent way of converting slices to list and oposite.

Reply:
Indeed. I returned the functions to `teal`, there are now copies in both packages.

[naming issue here](https://github.com/insightsengineering/teal.slice/issues/382)